lib/common/xml.c

```diff
@@ -1809,45 +1816,47 @@ find_xml_node(xmlNode * root, const char *search_path, gboolean must_find)
     return NULL;
 }
 
-/* As the name suggests, the perfect match is required for both node
-   name and fully specified attribute, otherwise, when attribute not
-   specified, the outcome is the first node matching on the name. */
-static xmlNode *
-find_entity_by_attr_or_just_name(xmlNode *parent, const char *node_name,
-                                 const char *attr_n, const char *attr_v)
-{
-    xmlNode *child;
+#define attr_matches(c, n, v) pcmk__str_eq(crm_element_value((c), (n)), \
+                                           (v), pcmk__str_none)
 
+/*!
+ * \internal
+ * \brief Find first XML child element matching given criteria
+ *
+ * \param[in] parent     XML element to search
+ * \param[in] node_name  If not NULL, only match children of this type
+ * \param[in] attr_n     If not NULL, only match children with an attribute
+ *                       of this name and a value of \p attr_v
+ * \param[in] attr_v     If \p attr_n and this are not NULL, only match children
+ *                       with an attribute named \p attr_n and this value
+ *
+ * \return Matching XML child element, or NULL if none found
+ */
+xmlNode *
+pcmk__xe_match(xmlNode *parent, const char *node_name,
+               const char *attr_n, const char *attr_v)
+{
     /* ensure attr_v specified when attr_n is */
     CRM_CHECK(attr_n == NULL || attr_v != NULL, return NULL);
 
-    for (child = __xml_first_child(parent); child != NULL; child = __xml_next(child)) {
-        /* XXX uncertain if the first check is strictly necessary here */
-        if (pcmk__str_eq(node_name, (const char *)child->name, pcmk__str_null_matches)) {
-            if (attr_n == NULL
-                    || pcmk__str_eq(crm_element_value(child, attr_n), attr_v, pcmk__str_none)) {
-                return child;
-            }
+    for (xmlNode *child = __xml_first_child(parent); child != NULL;
+         child = __xml_next(child)) {
+        if (pcmk__str_eq(node_name, (const char *) (child->name),
+                         pcmk__str_null_matches)
+            && ((attr_n == NULL) || attr_matches(child, attr_n, attr_v))) {
+            return child;
         }
     }
-
-    crm_trace("node <%s%s%s%s%s> not found in %s", crm_str(node_name),
-              attr_n ? " " : "",
-              attr_n ? attr_n : "",
-              attr_n ? "=" : "",
-              attr_n ? attr_v : "",
+    crm_trace("XML child node <%s%s%s%s%s> not found in %s",
+              (node_name? node_name : "(any)"),
+              (attr_n? " " : ""),
+              (attr_n? attr_n : ""),
```

Comment:
The closest I know of is

    #  define crm_str(x)    (const char*)(x?x:"<null>")

Maybe something like

    #define pcmk__str_else(s, d)  (s? s : d)
    #define pcmk__space_if(s)  (s? " " : "")

so it looks like

    crm_trace("XML child node <%s%s%s%s%s> not found in %s",
                     pcmk__str_else(node_name, "(any)"),
                     pcmk__space_if(attr_n),
                     pcmk__str_else(attr_n, ""),

It's not necessarily shorter but borderline more readable. Not sure if it's worth having to remember them (or run cocci) though.

Reply:
Ah yeah, `crm_str` is what I was thinking of.  Those are definitely not shorter, and are perhaps of dubious utility.    I just couldn't help but notice how often `s ? s : ""` is being used, and that plus your previous patches that got rid of a lot of uses of the ternary operator made it seem like a good next step.

Reply:
We specifically do "space and string if string is non-NULL otherwise empty strings" a lot, so maybe something like:

    #define PCMK__STR_FMT "%s%s"
    #define PCMK__STR_ARGS(s) ((s)? " " : ""), ((s)? (s) : "")

so it becomes

    crm_trace("XML child node <%s" PCMK__STR_FMT "%s%s> not found in %s",
                 (node_name? node_name : "(any)"),
                PCMK__STR_ARGS(attr_n),

though that's not exactly intuitive.

Reply:
Whew, yeah I don't think that's better.